README.md
# poly-jsbridge-sdk

[![CI status][github-action-image]][github-action-url]
[![codecov][codecov-image]][codecov-url]
[![NPM version][npm-image]][npm-url]
[![NODE version][node-image]][node-url]

[github-action-image]: https://github.com/hsycc/poly-jsbridge-sdk/workflows/test/badge.svg
[github-action-url]: https://github.com/hsycc/poly-jsbridge-sdk/actions/workflows/test.yml
[codecov-image]: https://coveralls.io/repos/github/hsycc/poly-jsbridge-sdk/badge.svg?branch=main
[codecov-url]: https://coveralls.io/github/hsycc/poly-jsbridge-sdk?branch=main
[npm-image]: https://img.shields.io/badge/npm-v4.16.13-blue
[npm-url]: http://npmjs.org/package/poly-jsbridge-sdk

[node-image]: https://img.shields.io/badge/node-v12.0.0-blue
[node-url]: https://nodejs.org/

Poly jsbridge protocol scheme.

Support with JavascriptChannel and navigationDelegate (url scheme).

Poly App 的 jsbridge 协议方案。

支持 JavascriptChannel 和 navigationDelegate。

## Install

```bash
npm install poly-jsbridge-sdk
```

```bash
yarn add poly-jsbridge-sdk
```

## Config

mode: JavascriptChannel

example:

window.postMessage(msg) || window[JavascriptChannelName].postMessage(msg)

msg format:

{"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"}

```ts
import { Config, SendModeEnum } from 'poly-jsbridge-sdk';

const initConfig: Config = {
  maxTryTimes: 3,
  isDebug: true,
  protocol: '',
  mode: SendModeEnum.CHANNEL, //  'channel' | 'url'
  JavascriptChannelName: 'PolySdk', // your app JavascriptChannel name,  can use 'window'
};
```

mode: navigationDelegate (url scheme)

example:

document.createElement('IFRAME');
iframe.setAttribute('src', url);

url scheme:

encodeURIComponent('poly://toast?payload={"message":"say hello"}?callbackId=poly_sdk_callback_163699755451280')

<!-- poly://toast?payload=%7B%22message%22%3A%22say%20hello%22%7D?callbackId=poly_sdk_callback_163699755451280 -->

```ts
import { Config, SendModeEnum } from 'poly-jsbridge-sdk';

const initConfig: Config = {
  maxTryTimes: 3,
  isDebug: true,
  protocol: 'poly://', // your custom protocol
  mode: SendModeEnum.URL,
  JavascriptChannelName: '',
};
```

## Usage

in cjs/esm

```ts
import { polyJsbridgeSdk, ClientResponse } from 'poly-jsbridge-sdk';

const polyJsbridgeSdk: PolyJsbridgeSdk = new PolyJsbridgeSdk(initConfig);

const toast = async () => {
  try {
    // send data to app
    const res: ClientResponse = await polyJsbridgeSdk.toast('你好', true);
    // res.code === 200
  } catch (error) {
    // res.code !== 200 or error
    console.log(error?.message);
  }
};

toast();

// NOTE:
// polyJsbridgeSdk.toast('say hello', true); be equivalent to
// polyJsbridgeSdk.$jsbridge.$call(
//   'toast',
//   {
//     message,
//   },
//   hasCallback // has ackcallback web to once ack
//   // 是否回调web完成一次ack
// );

// the app will received this data. 
// (example: {"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"})
// app 将接收这些数据。 
// (example: {"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"})
//
// after format this data and process the corresponding logic in app, response your data. 
// response data  needs to be followed ClientResponse: '{code: number, data: any ,  message: string;}'
// 格式化该数据后，在 app 中处理相应的逻辑, 回传响应结果,
// 响应数据需要遵循 ClientResponse 格式 {code: number, data: any ,  message: string;}
//
// in app webController
// await webController.evaluateJavascript("window.$jsbridge.$poly_sdk_callback_1632727252090577(${your response data});");
// be equivalent to window.eval in web
//
// window.$jsbridge.$poly_sdk_callback_1632727252090577 will be destroyed after once call.
// window.$jsbridge.$poly_sdk_callback_1632727252090577 将在一次调用后销毁。

```

in browser

```html
<!DOCTYPE html>
<html>
  <head>
    <script src=".../dist/browser/polyJsbridgeSdk.js"></script>
  </head>
  <body>
    <script>
      const polyJsbridgeSdk = new PolyJsbridgeSdk.default(initConfig);
      // ... same as in cjs/esm
    </script>
  </body>
</html>
```

## jsbridge registers function

The web actively registers for the app to call.

web 主动注册函数 让 app 调用。

in web

```ts
// after init PolyJsbridgeSdk ||  init Jsbridge
// 初始化 PolyJsbridgeSdk 或者 Jsbridge 后
$jsbridge.$register('log2');
$jsbridge.$on('log2', (result: any) => {
  console.log(result);
});

// you can see the function has bound in $jsbridge。
console.log($jsbridge.log2);

// you can off this function.
// 可以注销主动注册的函数。
$jsbridge.$off('log2', () => {
  // you code
});

console.log($jsbridge.log2); // throw error
```

in app

```dart
await webController.evaluateJavascript("window.$jsbridge.$log2(${your log});");
// be equivalent to : in web
// 等价于在 web 环境执行
// await window.eval("window.$jsbridge.$log2(${your log});");
// then it's can trigger the function with jsbridge.$on bound in web.
// 触发 jsbridge.$on 绑定的函数。

```

## extends / 继承

```ts
  import { PolyJsbridgeSdk Config } from 'poly-jsbridge-sdk';
  class YourSDK extend PolyJsbridgeSdk {
    constructor(config?: Config) {
      super(config);
    }
    public async _init(): Promise<void> {
      super._init(); // extends PolyJsbridgeSdk.init

      // PolyJsbridgeSdk.init has do this.
      // @ts-ignore
      // window.$jsbridge = this.$jsbridge; // mount to a global variable 
      // this.$register('log');
      // this.$on('log', (result: any) => {
      //   console.log(result);
      // });

      // add you register function
      this.$register('log2');
      this.$on('log2', (result: any) => {
        console.log(result);
      });
      // ...
    }
  }

```

## app jsbridge principle / app 桥接实现

complete parallel example : see [poly-webview-flutter-jsbridge-demo](https://github.com/hsycc/poly-webview-flutter-jsbridge-demo)

```dart
import 'package:webview_flutter/webview_flutter.dart';


// after web  emit polyJsbridgeSdk.toast('say hello', true);
// in webviewCtrl
const webviewCtrl  = weview(
  // ...
  // mode javascriptChannel
  javascriptChannels: <JavascriptChannel>[
    JavascriptChannel(
      name: 'PolySdk',
      onMessageReceived: (JavascriptMessage message) async {
        String jsonStr = message.message;

        // serialization/序列化
        var model = parseJson(jsonStr); // {"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"}
        // add you code
        // const resultStr = ...
        await webController.evaluateJavascript("window.$jsbridge.$callbackId($resultStr);");

    }),
  ].toSet(),
  //  mode navigationDelegate (url scheme)
  navigationDelegate: (NavigationRequest request) {
      if (request.url.startsWith('poly://')) {
        // request.url poly://toast?payload=%7B%22message%22%3A%22say%20hello%22%7D?callbackId=poly_sdk_callback_163699755451280
        // serialization/序列化
        // var model = ... format request.url; // {"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"}
        // add you code
        // const resultStr ...
        await webController.evaluateJavascript("window.$jsbridge.$callbackId($resultStr);");

        return NavigationDecision.prevent; // reject navigate
      }
      return NavigationDecision.navigate; // narmal navigate other url
    },
  // ...
)


```

## TODO：

- [ ] eventEmitter use typescript
- [ ] fix @ts-ignore



CHANGELOG.md
# 1.0.0 (2021-11-18)


### Features

* add core code ([be67d45](https://github.com/hsycc/poly-jsbridge-brigde/commit/be67d452dc3c3043cfbeda01fdf4c15924ae1053))
* add webpack to build browser version ([46994e8](https://github.com/hsycc/poly-jsbridge-brigde/commit/46994e8262852216cc0a08cb849773c23329d43f))
* build changes ([5e7ebc0](https://github.com/hsycc/poly-jsbridge-brigde/commit/5e7ebc0e59913071d9480e76c545d9b7237503b8))
* rename package name , update  test cases, update doc ([658f438](https://github.com/hsycc/poly-jsbridge-brigde/commit/658f43805796d94c4f98b98c62d47b2c2199e01f))





lib/jsbridge.ts
/**
 * @format
 * @Author: hsycc
 * @Date: 2021-09-21 18:22:19
 * @LastEditTime: 2021-11-18 15:46:01
 * @Description:
 */

import { Config, ClientResponse, SendModeEnum } from './type';

import EventEmitter from 'events';

class Jsbridge {
  public config: Config;
  private tryTimes = 0;
  private isInBrowser: boolean;
  private event: any;
  constructor(config: Config) {
    this.event = new EventEmitter();
    this.config = config;
    this.isInBrowser = !!(typeof window !== 'undefined');
  }

  /** call message and send to client */
  public $call(
    method: string,
    payload: Record<string, unknown>,
    hasCallback: boolean | string,
  ): Promise<ClientResponse> {
    payload = payload || {};
    // in browser
    if (this.isInBrowser) {
      // TODO: custom callback id case calls the registered function
      // if (typeof hasCallback === 'string') {
      // }

      const that = this;

      return new Promise((resolve, reject) => {
        const callbackId = hasCallback ? this.getCallbackId() : '';
        const message = this.generateMessage(method, payload, callbackId);
        if (that.config.isDebug) {
          console.debug('send', message);
        }

        if (callbackId) {
          this.$register(callbackId);
          this.$on(callbackId, (result: ClientResponse) => {
            if (that.config.isDebug) {
              console.debug(`[${callbackId}]:`, result);
            }
            try {
              if (result.code === 200) {
                resolve(result);
              } else {
                // status code error
                reject(result);
              }
            } catch (error) {
              reject(error);
            } finally {
              // unsubscribe callbackId
              this.$off(callbackId);
            }
          });
        }

        this.sendMessage(message);
      });
    } else {
      console.error('does not support poly-jsbridge-sdk');
      return new Promise((resolve, reject) => {
        reject('does not support poly-jsbridge-sdk');
      });
    }
  }

  /** subscribe */
  public $on(name: string, func: (result: ClientResponse) => any): any {
    if (this.event._events[name]) {
      this.$off(name);
    }
    this.event.on(name, func);
    return this;
  }

  /** unsubscribe */
  public $off(name: string, func?: (result: ClientResponse) => any): any {
    if (func) {
      this.event.removeListener(name, func);
    } else {
      this.event.removeAllListeners(name);
    }
    return this;
  }

  /** register */
  public $register(name: string): void {
    const ar: string[] = name.split('.');
    const len: number = ar.length;
    let obj: any = this;
    const eventName = ar.join('_');

    ar.forEach((el, idx) => {
      if (idx === len - 1) {
        // @ts-ignore
        obj[el] = (json: Record<string, unknown>) => {
          this.event.emit(eventName, json);
        };
      } else {
        // @ts-ignore
        obj[el] = obj[el] || {};
        // @ts-ignore
        obj = obj[el];
      }
    });
  }

  /**
   * Generate the message
   *   {"method":"toast","payload":{"message":"你好"},"callbackId":"poly_sdk_callback_1632727252090577"}
   *   or
   *   poly://toast?payload=%7B%22message%22%3A%22say%20hello%22%7D?callbackId=poly_sdk_callback_163699755451280
   */
  private generateMessage(method: string, payload: Record<string, unknown>, callbackId?: string): string {
    if (this.config.mode === SendModeEnum.CHANNEL) {
      const message = {
        method,
        payload,
      };

      if (callbackId) {
        Object.assign(message, {
          callbackId,
        });
      }

      return JSON.stringify(message);
    } else {
      const payloadData = this.encode(payload);
      let message = this.config.protocol.concat(method).concat('?payload=').concat(payloadData);

      if (callbackId) {
        message = message.concat('&callbackId=').concat(callbackId);
      }

      return message;
    }
  }

  /**
   *
   * encode the data
   * Parse Javascript Object to params String
   * `JSON.stringify() -> encodeURIComponent()`
   */
  private encode(obj: Record<string, unknown>): string {
    const json = JSON.stringify(obj);
    return encodeURIComponent(json);
  }

  /**
   *
   * decode the data
   * Parse url params data to string
   * `decodeURIComponent() -> JSON.parse()`
   */
  private decode(str: string): Record<string, unknown> {
    // const decodeURL = atob(str);
    const jsonStr = decodeURIComponent(str);
    return JSON.parse(jsonStr);
  }

  /**
   * Generate random string, begin with `poly_sdk_callback_[random]`
   */
  private getCallbackId(): string {
    const random = parseInt(Math.random() * 10000 + '');
    return 'poly_sdk_callback_' + new Date().getTime() + random;
  }

  /**
   * Post message to client
   */
  public sendMessage(msg: string): void {
    // browser
    if (this.isInBrowser) {
      try {
        if (this.config.mode === SendModeEnum.CHANNEL) {
          // @ts-ignore
          this.config.JavascriptChannelName === 'window'
            ? // @ts-ignore
              window.postMessage(msg)
            : // @ts-ignore
              window[this.config.JavascriptChannelName].postMessage(msg);
        } else {
          // create iframe
          // const document = window.document
          let iframe: HTMLElement | null = document.createElement('IFRAME');
          iframe.setAttribute('src', msg);
          document.documentElement.appendChild(iframe);
          iframe.parentNode?.removeChild(iframe);
          iframe = null;
        }
        // reset the tryTime to first
        this.tryTimes = 1;
      } catch (error) {
        // per second will attempt to reinvoke, failed over  maxTryTimes will never reinvoked again
        if (this.tryTimes < this.config.maxTryTimes) {
          setTimeout(() => {
            this.sendMessage(msg);
            this.tryTimes = ++this.tryTimes;
          }, 1000);
        } else {
          throw new Error(`post msg timeout( ${this.config.maxTryTimes} times): ${msg} \n , error: ${error}`);
        }
      }
    } else {
      // nodejs
      throw new Error('Method not implemented.');
    }
  }
}

export default Jsbridge;
